### PR TITLE
Backport replay additions from pulumi-terraform-bridge

### PR DESCRIPTION
**Pull request by t0yv0**
Alternative to https://github.com/pulumi/providertest/pull/59

The motivation is to deprecate the replay module in pulumi-terraform-bridge in favor of a Go package here under generic providertest, and backport edits that happened in pulumi-terraform-bridge to this new home.

Unlike #59 two small things:

- trying not to expose the match options but solve the CheckFailure ordering on semantic level
- trying to make Errors compatible with gRPC log capture format []string

### CODE REVIEW

File: replay/replay.go [hunk 1]
```diff
@@ -232,8 +269,44 @@ func ReplayFile(t *testing.T, server pulumirpc.ResourceProviderServer, traceFile
 	assert.Greater(t, count, 0)
 }
 
+// See also: https://github.com/pulumi/pulumi/blob/master/pkg/util/rpcdebug/logformat.go#L28
 type jsonLogEntry struct {
 	Method   string          `json:"method"`
 	Request  json.RawMessage `json:"request,omitempty"`
 	Response json.RawMessage `json:"response,omitempty"`
+	Errors   []string        `json:"errors,omitempty"`
```
Comment (t0yv0) [inline replay/replay.go:277]: https://github.com/pulumi/pulumi-terraform-bridge/pull/1496/ introduced Errors as *string but I think that's a problem because GRPC logs capture them as []string. So it inhibits grabbing GRPC logs and reusing them as asserts. Trying to fixup  here. Let me see that the tests from 1496 can still run with these changes.

Comment (t0yv0) [reply replay/replay.go:277]: Yes it looks like with minor changes muxer_test can still work fine: https://github.com/pulumi/pulumi-terraform-bridge/pull/1638

File: replay/replay.go [hunk 2]
```diff
@@ -94,45 +95,45 @@ func Replay(t *testing.T, server pulumirpc.ResourceProviderServer, jsonLog strin
...
-		replay(t, entry, new(pulumirpc.CallRequest), server.Call)
+		replay(t, entry, new(pulumirpc.CallRequest), server.Call, normCallResponse)
 
 	case "/pulumirpc.ResourceProvider/Check":
-		replay(t, entry, new(pulumirpc.CheckRequest), server.Check)
+		replay(t, entry, new(pulumirpc.CheckRequest), server.Check, normCheckResponse)
 
 	case "/pulumirpc.ResourceProvider/Diff":
-		replay(t, entry, new(pulumirpc.DiffRequest), server.Diff)
+		replay(t, entry, new(pulumirpc.DiffRequest), server.Diff, nil)
```
Comment (iwahbe) [inline replay/replay.go:121]: This gives us an excellent entry point to sort `stables`. 😄

That should probably be a different PR.

**Review by iwahbe — APPROVED** [review]
This is slightly cleaner than #59, since it takes advantage of `replay` being specific to Pulumi gRPC.

File: replay/replay.go (hunk 1 above)
Comment (iwahbe) [reply replay/replay.go:277]: This PR contains an invalid go.mod. It is not demonstrative.

Comment (t0yv0) [reply replay/replay.go:277]: I might have screwed up something with go workspaces. But I'm pretty sure the tests can be made to pass with some small changes. If not we can always iterate.

Comment (iwahbe) [reply replay/replay.go:277]: I want to make sure that we can make tests pass before we merge. I'm comfortable merging with you saying "The tests have passed locally". We don't need to public facing CI here.

Comment (t0yv0) [reply replay/replay.go:277]: I'm sure you overestimate the public interest in reading our PRs :)
